haskell/test/Test/OffChainView.hs

```diff
@@ -0,0 +1,80 @@
+-- SPDX-FileCopyrightText: 2021 Arthur Breitman
+-- SPDX-License-Identifier: LicenseRef-MIT-Arthur-Breitman
+
+module Test.OffChainView
+  ( test_OffChainViews
+  ) where
+
+import Universum hiding (view)
+
+import Fmt (pretty)
+
+import Lorentz.Value
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Lorentz.Contracts.Spec.TZIP16Interface (View)
+import Michelson.Interpret (MichelsonFailed)
+import Michelson.Test.Dummy
+import Morley.Metadata
+import Test.HUnit ((@?=))
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (testCase)
+import Tezos.Address
+
+import SegCFMM.Types
+import SegCFMM.TZIP16Metadata
+import Test.SegCFMM.Storage (defaultStorage)
+
+xToken :: Address
+xToken = unsafeParseAddress "KT1XMaGzMc7iGVkbnCC9gZyjs2b85A3NxJTz"
+
+yToken :: Address
+yToken = unsafeParseAddress "KT1N3SQzqEb5vU7HmtqSqnuSTgrxs8hwkqVU"
+
+xTokenId :: FA2.TokenId
+xTokenId = FA2.TokenId 0
+
+yTokenId :: FA2.TokenId
+yTokenId = FA2.TokenId 1
+
+offChainViewStorage :: Storage
+offChainViewStorage =
+  defaultStorage
+    { sConstants = (sConstants defaultStorage)
+      { cXTokenAddress = xToken
+      , cXTokenId = xTokenId
+      , cYTokenAddress = yToken
+      , cYTokenId = yTokenId
+      }
+    }
+
+test_OffChainViews :: TestTree
+test_OffChainViews =
+  let
+    mc = mkMetadataSettings defaultMetadataConfig
+  in
+    testGroup "Off-chain views"
+    [ testCase "Get the address of tokens X" $
+        runView @Address (getTokenXAddressView mc) offChainViewStorage NoParam
+          @?= (Right $ xToken)
+    , testCase "Get the address of tokens Y" $
+        runView @Address (getTokenYAddressView mc) offChainViewStorage NoParam
+          @?= (Right $ yToken)
+    , testCase "Get the id of tokens X" $
+        runView @FA2.TokenId (getTokenXIdView mc) offChainViewStorage NoParam
+          @?= (Right $ xTokenId)
+    , testCase "Get the id of tokens Y" $
+        runView @FA2.TokenId (getTokenYIdView mc) offChainViewStorage NoParam
+          @?= (Right $ yTokenId)
+    ]
+
+runView
+  :: forall ret. (HasCallStack, IsoValue ret)
+  => View $ ToT Storage
+  -> Storage
+  -> ViewParam
+  -> Either MichelsonFailed ret
+runView view storage param =
+  case interpretView dummyContractEnv view param storage of
+    Right x -> Right x
+    Left (VIEMichelson _ (MSVIEMichelsonFailed e)) -> Left e
+    Left err -> error (pretty err)
```

Comment:
I used this function form BaseDAO: https://github.com/tezos-commons/baseDAO/blob/b3aa7886950d4f1eb65816ed726ce69e77e14472/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs#L46-L56

I believe when it return `Left VIEMichelson`, we would want to compare the error to what we expect. But otherwise we throw the runtime error. I don't think we have a used case in yet, but I just want to keep it the way it is.

Reply:
> we would want to compare the error to what we expect

It's not really clear why, but ok.

Still, since we're using hunit here, we should use hunit's [`assertFailure`](https://hackage.haskell.org/package/tasty-hunit-0.10.0.3/docs/Test-Tasty-HUnit.html#v:assertFailure) and not `error`.

Reply:
Got this, I've updated it.